### PR TITLE
KAFKA-8884: class cast exception improvement

### PR DESCRIPTION
**Pull request by astubbs**
If a processor causes a class cast exception, atm you get a bit of a cryptic error if you're not used to them, and without a context sensitive suggestion as to what could be wrong. Often these can be cause by miss-configured Serdes (defaults).

Old error:
```
Caused by: java.lang.ClassCastException: class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
```

An example of the improvement over the case exception:
```
org.apache.kafka.streams.errors.StreamsException: Exception caught in process. taskId=0_0, processor=KSTREAM-SOURCE-0000000000, topic=streams-plaintext-input, partition=0, offset=0, stacktrace=org.apache.kafka.streams.errors.StreamsException: 
ClassCastException invoking Processor. Do the Processor's input types match the deserialized types? Check 
the Serde setup and change the default Serdes in StreamConfig or provide correct Serdes via method 
parameters. Deserialized input types are: key: [B, value: [B
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:123)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:201)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:180)
	at org.apache.kafka.streams.processor.internals.ProcessorContextImpl.forward(ProcessorContextImpl.java:133)
	at org.apache.kafka.streams.processor.internals.SourceNode.process(SourceNode.java:87)
	at org.apache.kafka.streams.processor.internals.StreamTask.process(StreamTask.java:366)
	at org.apache.kafka.streams.TopologyTestDriver.pipeInput(TopologyTestDriver.java:419)
	at org.apache.kafka.streams.processor.internals.ProcessorNodeTest.testTopologyLevelClassCastException(ProcessorNodeTest.java:176)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:365)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:330)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:78)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:328)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:65)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:292)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:305)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:412)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.ClassCastException: class [B cannot be cast to class java.lang.String ([B and java.lang.String are in module java.base of loader 'bootstrap')
	at org.apache.kafka.streams.kstream.internals.AbstractStream.lambda$withKey$1(AbstractStream.java:103)
	at org.apache.kafka.streams.kstream.internals.KStreamFlatMapValues$KStreamFlatMapValuesProcessor.process(KStreamFlatMapValues.java:40)
	at org.apache.kafka.streams.processor.internals.ProcessorNode.process(ProcessorNode.java:119)
	... 32 more
```

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (astubbs) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]:
```suggestion
```

**Comment by mjsax** [comment]
@astubbs `:streams:checkstyleTest` failed -- can you address it and update the PR?

**Comment by vvcephei** [comment]
Hey @astubbs , this looks good to me, mod the failing build of course.

The only question I had is whether the error message should indicate that the cast exception _might_ have another cause (in user code, for example), but that wrong serdes are a common cause of the error. For example, if a processor wires in a store, but casts the generics incorrectly, we would also raise a class cast exception during processing, but the cause would not be wrong serdes.

WDYT?

**Comment by astubbs** [comment]
Do you want both the full topology tests?
I’ll also take a stab at adding a test for the bad store setup?

**Comment by vvcephei** [comment]
Hey @astubbs , Sorry for the delay.

I'm not sure if we need any more tests, I was just asking about the wording of the exception message itself.

**Comment by astubbs** [comment]
please retest

**Comment by astubbs** [comment]
Please test again

**Comment by astubbs** [comment]
Retest this please

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java]:
```suggestion
                    + "the generics incorrectly, a class cast exception could be raised during processing, but the "
```

**Review by vvcephei — APPROVED** [review]
LGTM, @astubbs ! Just one typo.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java
Comment (astubbs) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java:96]: Add unit test for this?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java:96]: Not sure what you mean?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java]: typo: `also although`

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: As we catch the exception we should swallow is and don't expect it

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: `ConsumerRecordFactory` is deprecated via KIP-470 (merge recently) -- please update the code to use `TestInputTopic` -- we cannot merge as-is, because the build would fail if we use deprecated classes.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: remove try-catch and replace with:
```
final StreamsException s = assertThrows(StreamsException.class, () -> testDriver.pipeInput(consumerRecord));
```
assert afterwards and don't re-throw.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: Do we need this test?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: rewrite test as above using `assertThrows()`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: Why do we need to suppress warnings?

**Comment by mjsax** [comment]
Overall LGTM -- some testing nits and we need to rewrite before merging to not break the build.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java
Comment (astubbs) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorNode.java]:
```suggestion
                    + "Note that although incorrect Serdes are a common cause of error, the cast exception might have "
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java
Comment (astubbs) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java:96]: Sorry, I was trying to ask if I should write a unit test to capture the scenarios I was describing in the highlighted code - a test to see if the exception is used or not when using Produced.keySerde method.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (astubbs) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: But the catch is only there to test the message content. If the exception is incorrectly not thrown, the test would incorrectly pass. 
Ah, I see below, you're saying the preferred style is to use using assertThrows(), unless the exception is not needed to be caught explicitly in the test?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (astubbs) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: Will check it out.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (astubbs) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: Yeah I was thinking about this, I don't think we do. I'll remove it.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (astubbs) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: I'll have another look, thanks...

**Review by astubbs — COMMENTED** [review]
Thanks for the review!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java
Comment (astubbs) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorNodeTest.java]: Oh man sooo much nicer. I was reading up about it and talking to the author but actually using it is soo much nicer!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java:96]: I think the existing test are sufficient.

**Comment by mjsax** [comment]
Merged to `trunk` and cherry-picked to `2.4`.

Thanks for the PR @astubbs!

**Pull request merged**
